examples_tests/49.NewAPITest/main.cpp

```diff
@@ -0,0 +1,330 @@
+		assert(rpindependent_pipeline);
+	}
+
+	core::smart_refctd_ptr<video::IGPUGraphicsPipeline> pipeline;
+	{
+		video::IGPUGraphicsPipeline::SCreationParams gp_params;
+		gp_params.rasterizationSamplesHint = asset::IImage::ESCF_1_BIT;
+		gp_params.renderpass = renderpass;
+		gp_params.renderpassIndependent = rpindependent_pipeline;
+		gp_params.subpassIx = 0u;
+
+		pipeline = device->createGPUGraphicsPipeline(nullptr, std::move(gp_params));
+	}
+
+	core::smart_refctd_ptr<video::IGPUBuffer> buffer;
+	{
+		const SVertex vertices[3]{
+			{{0.0f, -0.5f}, {1.0f, 0.0f, 0.0f}},
+			{{0.5f, 0.5f}, {0.0f, 1.0f, 0.0f}},
+			{{-0.5f, 0.5f}, {0.0f, 0.0f, 1.0f}}
+		};
+
+		video::IDriverMemoryBacked::SDriverMemoryRequirements mreq;
+		
+		buffer = device->createDeviceLocalGPUBufferOnDedMem(sizeof(vertices));
+		assert(buffer);
+
+		core::smart_refctd_ptr<video::IGPUCommandBuffer> cb;
+		device->createCommandBuffers(cmdpool.get(), video::IGPUCommandBuffer::EL_PRIMARY, 1u, &cb);
+		assert(cb);
+
+		cb->begin(video::IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+		cb->updateBuffer(buffer.get(), 0u, sizeof(vertices), vertices);
+		cb->end();
+		
+		video::IGPUQueue::SSubmitInfo info;
+		auto* cb_ = dynamic_cast<video::IGPUPrimaryCommandBuffer*>(cb.get());
+		info.commandBufferCount = 1u;
+		info.commandBuffers = &cb_;
+		info.pSignalSemaphores = nullptr;
+		info.signalSemaphoreCount = 0u;
+		info.pWaitSemaphores = nullptr;
+		info.waitSemaphoreCount = 0u;
+		info.pWaitDstStageMask = nullptr;
+		queue->submit(1u, &info, nullptr);
+	}
+
+	core::smart_refctd_ptr<video::IGPUCommandBuffer> cmdbuf[SC_IMG_COUNT];
+	device->createCommandBuffers(cmdpool.get(), video::IGPUCommandBuffer::EL_PRIMARY, SC_IMG_COUNT, cmdbuf);
+	for (uint32_t i = 0u; i < SC_IMG_COUNT; ++i)
+	{
+		auto& cb = cmdbuf[i];
+		auto& fb = fbo[i];
+
+		cb->begin(0);
+		
+		auto* buf = buffer.get();
+		size_t offset = 0u;
+		cb->bindVertexBuffers(0u, 1u, &buf, &offset);
+		cb->bindGraphicsPipeline(pipeline.get());
+		video::IGPUCommandBuffer::SRenderpassBeginInfo info;
+		asset::SClearValue clear;
+		asset::VkRect2D area;
+		area.offset = { 0, 0 };
+		area.extent = { WIN_W, WIN_H };
+		clear.color.float32[0] = 1.f;
+		clear.color.float32[1] = 0.f;
+		clear.color.float32[2] = 0.f;
+		clear.color.float32[0] = 1.f;
+		info.renderpass = renderpass;
+		info.framebuffer = fb;
+		info.clearValueCount = 1u;
+		info.clearValues = &clear;
+		info.renderArea = area;
+		cb->beginRenderPass(&info, asset::ESC_INLINE);
+		cb->draw(3u, 1u, 0u, 0u);
+		cb->endRenderPass();
+
+		cb->end();
+	}
+
+	constexpr uint32_t FRAME_COUNT = 5000u;
+	constexpr uint64_t MAX_TIMEOUT = 99999999999999ull; //ns
+	for (uint32_t i = 0u; i < FRAME_COUNT; ++i)
+	{
+		auto img_acq_sem = device->createSemaphore();
+		auto render_finished_sem = device->createSemaphore();
+
+		uint32_t imgnum = 0u;
+		sc->acquireNextImage(MAX_TIMEOUT, img_acq_sem.get(), nullptr, &imgnum);
```

Comment:
this `acquire` does not officially guarantee that the commandbuffer we want to submit has finished executing, and unless it has the following flag:
> VK_COMMAND_BUFFER_USAGE_SIMULTANEOUS_USE_BIT specifies that a command buffer can be resubmitted to a queue while it is in the pending state, and recorded into multiple primary command buffers.
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkCommandBufferUsageFlagBits.html 

You could run into a situation where you're trying to submit a commandbuffer that is currently pending or being executed on the GPU, which is a no-no without that flag.

Ergo I recommend creating N fences in the signalled state, waiting on the fence just before the submit on line 312, resetting it and then signalling as part of the submit.

Reply:
> this acquire does not officially guarantee that the commandbuffer we want to submit has finished executing, and unless it has the following flag

it guarantees that the acquired image has already been presented. In order to present a frame, all rendering for this frame must be finished, so i think it guarantees that the cmdbuf has finished

Reply:
from the Vulkan spec
> The presentation engine may not have finished reading from the image at the time it is acquired, so the application must use semaphore and/or fence to ensure that the image layout and contents are not modified until the presentation engine reads have completed.

Reply:
Multi-buffer commandpools and commandbuffers (at least one pool per frame) just like the new ex 29

@Devsh-Graphics-Programming/mids @Devsh-Graphics-Programming/seniors @achalpandeyy